zkevm-circuits/src/evm_circuit/execution/signed_comparator.rs

```diff
@@ -0,0 +1,267 @@
+use halo2::{arithmetic::FieldExt, circuit::Region, plonk::Error};
+
+/// Gadget that implements the ExecutionGadget trait to handle the Opcodes SLT
+/// and SGT.
+#[derive(Clone, Debug)]
+pub(crate) struct SignedComparatorGadget<F> {
+    same_context: SameContextGadget<F>,
+
+    a: Word<F>,
+    b: Word<F>,
+
+    sign_check_a: LtGadget<F, 1>,
+    sign_check_b: LtGadget<F, 1>,
+    lt_lo: LtGadget<F, 16>,
+    lt_hi: LtGadget<F, 16>,
+
+    is_sgt: IsEqualGadget<F>,
+}
+
+impl<F: FieldExt> ExecutionGadget<F> for SignedComparatorGadget<F> {
+    const NAME: &'static str = "SCMP";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::SCMP;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+
+        let a = cb.query_word();
+        let b = cb.query_word();
+
+        // The Signed Comparator gadget is used for both opcodes [SLT]() and
+        // [SGT](). Depending on whether the opcode is SLT or SGT, we
+        // swap the order in which the inputs are placed on the stack.
+        let is_sgt =
+            IsEqualGadget::construct(cb, opcode.expr(), OpcodeId::SGT.expr());
+
+        // Both `a` and `b` are to be treated as two's complement signed 256-bit
+        // (32 cells) integers. This means, the first bit denotes the sign
+        // of the absolute value in the rest of the 255 bits. This means the
+        // number is negative if the most significant cell >= 128
+        // (0b10000000).
+        let sign_check_a =
+            LtGadget::construct(cb, a.cells[0].expr(), 128.expr());
+        let sign_check_b =
+            LtGadget::construct(cb, b.cells[0].expr(), 128.expr());
+
+        // sign_check_a_lt expression implies a is positive since its MSB < 2**7
+        // sign_check_b_lt expression implies b is positive since its MSB < 2**7
+        let sign_check_a_lt = sign_check_a.expr();
+        let sign_check_b_lt = sign_check_b.expr();
+
+        // We require the comparison check only for the cases where:
+        // (a < 0 && b < 0) || (a > 0 && b > 0).
+        // We ignore the equality expression since we only care about the LT
+        // check.
+        let lt_lo = LtGadget::construct(
+            cb,
+            from_bytes::expr(&a.cells[0..16]),
+            from_bytes::expr(&b.cells[0..16]),
+        );
+        let lt_hi = LtGadget::construct(
+            cb,
+            from_bytes::expr(&a.cells[16..32]),
+            from_bytes::expr(&b.cells[16..32]),
+        );
+        let a_lt_b_lo = lt_lo.expr();
+        let a_lt_b_hi = lt_hi.expr();
+
+        // If a and b both are positive, we do a standard comparison check. If
+        // both a and b are negative, we invert those outcomes.
+        let a_b_positive = sign_check_a_lt.clone() * sign_check_b_lt.clone();
+        let a_lt_b = select::expr(
+            a_b_positive,
+            select::expr(a_lt_b_hi.clone(), 1.expr(), a_lt_b_lo.clone()),
```

Comment:
You cannot directly use `a_lt_b_lo.clone()` when `a_lt_b_hi == 0`, and instead you need to add `a_hi == b_hi` check here.